src/net.cpp

```diff
@@ -1945,14 +1927,25 @@ void CConnman::ThreadOpenConnections(const std::vector<std::string> connect)
                 LogPrint(BCLog::NET, "Making feeler connection to %s\n", addrConnect.ToString());
             }
 
-            // Open this connection as block-relay-only if we're already at our
-            // full-relay capacity, but not yet at our block-relay peer limit.
-            // (It should not be possible for fFeeler to be set if we're not
-            // also at our block-relay peer limit, but check against that as
-            // well for sanity.)
-            bool block_relay_only = nOutboundBlockRelay < m_max_outbound_block_relay && !fFeeler && nOutboundFullRelay >= m_max_outbound_full_relay;
+            ConnectionType conn_type;
+            // Determine what type of connection to open. Open OUTBOUND
+            // connections until we meet our full-relay capacity, or if
+            // GetTryNewOutboundPeer() is true. If full-relay capacity is met,
+            // open BLOCK_RELAY connections until we hit our block-relay peer
+            // limit.
+            if (fFeeler) {
+                conn_type = ConnectionType::FEELER;
+            } else if (nOutboundFullRelay < m_max_outbound_full_relay) {
+                conn_type = ConnectionType::OUTBOUND;
+            } else if (nOutboundBlockRelay < m_max_outbound_block_relay) {
+                conn_type = ConnectionType::BLOCK_RELAY;
+            } else if (GetTryNewOutboundPeer()) {
+                conn_type = ConnectionType::OUTBOUND;
+            } else {
+                assert(false);
```

Comment:
I don't like this new assert. It seems safe for now because of the logic earlier in this function, but I don't think we should rely on that (or something else) not changing. Why not just maintain the old behaviour of this always creating `OUTBOUND` connections when they're not `FEELER` or `BLOCK_RELAY`

Reply:
hm, I don't quite follow your logic. If a developer were to change the function's logic flow without properly attending to this switch statement, in one case we'd silently open (short lived) `OUTBOUND` connections & in the other case the developer would hit a compiler error. is this different than how you're thinking about the use cases? 

also, see https://github.com/bitcoin/bitcoin/pull/19316#discussion_r446363611 for more convo on this topic

Reply:
This isn't a switch statement, and there aren't any compile time checks. `assert(false)` is a runtime error.

Here's what I'm worried about:

- in line 1856 (https://github.com/bitcoin/bitcoin/pull/19316/files#diff-9a82240fe7dfe86564178691cc57f2f1R1856), both outbound and block-relay-only are full, and `GetTryNewOutboundPeer()` is true, so we don't set `fFeeler` to true.
- meanwhile, on the message handler thread, `SetTryNewOutboundPeer(false)` is called.
- we then reach this line where `fFeeler` is false, both outbound and block-relay-only are full, and `GetTryNewOutboundPeer()` is false. We assert and terminate.

You've justified this change above:

> GetTryNewOutboundPeer is simply a wrapper for m_try_another_outbound_peer, which is set to true when a stale tip is detected (link).
> 
> so, when we get to this current conditional, if fFeeler if false & our full-relay and block-relay connection counts are at the allocated max -> GetTryNewOutboundPeer must be true.

`m_try_another_outbound_peer` is updated by another thread, and there's no lock on that data between the conditional above and this line, so there's no guarantee that it won't change. We may be lucky and some combination of other conditions in the eviction/stale tip logic in net_processing means this won't actually happen, but it seems possible that there's a window condition where it could, which would result in users' nodes very intermittently terminating.

I think this conn_type logic can probably be moved up and merged with the `fFeeler` setting logic above (`fFeeler`, `nOutboundFullRelay` and `nOutboundBlockRelay` don't get updated in the intervening code, so why not just set `conn_type` above?), but I don't think we should complicate this PR by doing that here. For this PR, I think we should just not add this assert.

Reply:
ah, I missed that its updated by a separate thread. thanks for this explanation!

Reply:
updated. nice catch!

Reply:
I've taken a stab at a further simplification on the follow-up PR #19724 here: 6bdc1e3a693081e1f12a1a0b2411ac129d2b2923

I've realized there are other circumstances under which the connection type defaults to `OUTBOUND`- since `fFeeler` uses a timer, it is possible for outbounds to be at capacity, block relays to be at capacity, `GetTryNewOutboundPeer` to be false, and `fFeeler` to be false, under which case we default to opening an `OUTBOUND` connection. 

very glad you prompted me to remove the assert! thank you!